custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ThingHandlerSetLabelCheck.java

```diff
@@ -0,0 +1,212 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.tools.analysis.checkstyle;
+
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.JAVA_EXTENSION;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Optional;
+import java.util.Queue;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openhab.tools.analysis.checkstyle.api.AbstractStaticCheck;
+
+import com.puppycrawl.tools.checkstyle.TreeWalker;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.FileText;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtils;
+
+import antlr.RecognitionException;
+import antlr.TokenStreamException;
+
+/**
+ * Checks the code and generates a warning if a class extends BaseThingHandler
+ * and the instance method setLabel of the Thing interface is used.
+ * The label have to be set only by the user.
+ *
+ * @author Tanya Georgieva - Initial Contribution
+ */
+public class ThingHandlerSetLabelCheck extends AbstractStaticCheck {
```

Comment:
I don't know about the performance of this but I personally doubt whether this check (with some corner cases which are extremely hard to cover) is worth iterating every file tree at least 1 additional time and every handler at least one additional time extra, @dstivanov wdyt?

@tanyatanyatanya Can you give some idea of the build time this would add to for example openHAB or ESH?

Reply:
@martinvw I discussed with Tanya that we will provide measurements for the build times without this check, with the initial changes and with the new ones (which requires more files processing). When this is in place we will have an objective argument which way to go. I still believe that the current version will not be so time-consuming as it is working on bundle level and filters out the files. Also, the increased build times recently are subject of our investigations currently (see here - https://github.com/openhab/static-code-analysis/issues/232) so I guess that soon they will be improved.

Reply:
But my main problem is that imho its not able (not even possible?) to cover all scenarios so that it even takes a limited time without providing real safety might be too much already.

Reply:
The build time without this check on openHAB is 17:09 minutes, with this new version of the check is 20:55 minutes and with the old version where I do not search for indirect inheritance is 18:45 minutes.